File: sdk/java/src/test/java/io/routeiq/sdk/HandlesTest.java
package io.routeiq.sdk;

import io.opentelemetry.sdk.testing.junit5.OpenTelemetryExtension;
import io.opentelemetry.sdk.trace.data.SpanData;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.RegisterExtension;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class HandlesTest {

    @RegisterExtension
    static final OpenTelemetryExtension otelTesting = OpenTelemetryExtension.create();

    private RouteIQ riq;

    @BeforeEach
    void setUp() {
        var sdk = (io.opentelemetry.sdk.OpenTelemetrySdk) otelTesting.getOpenTelemetry();
        riq = new RouteIQ("test-agent", sdk.getSdkTracerProvider());
    }

    private SpanData findSpan(String prefix) {
        return otelTesting.getSpans().stream()
                .filter(s -> s.getName().startsWith(prefix))
                .findFirst()
                .orElseThrow(() -> new AssertionError("No span with prefix: " + prefix));
    }

    private SpanData findSpanNamed(String name) {
        return otelTesting.getSpans().stream()
                .filter(s -> s.getName().equals(name))
                .findFirst()
                .orElseThrow(() -> new AssertionError("No span named: " + name));
    }

    private String attr(SpanData span, String key) {
        var sv = span.getAttributes().get(io.opentelemetry.api.common.AttributeKey.stringKey(key));
        if (sv != null) return sv;
        var lv = span.getAttributes().get(io.opentelemetry.api.common.AttributeKey.longKey(key));
        if (lv != null) return String.valueOf(lv);
        var dv = span.getAttributes().get(io.opentelemetry.api.common.AttributeKey.doubleKey(key));
        if (dv != null) return String.valueOf(dv);
        return null;
    }

    // ── TaskHandle ────────────────────────────────────────────────────────

    @Test
    void task_spanNameStartsWithTask() {
        try (var task = riq.task("find Paris")) { }
        assertNotNull(findSpan("task:"));
    }

    @Test
    void task_envelopeAttributes() {
        String taskId;
        try (var task = riq.task("find Paris")) {
            taskId = task.taskId;
        }
        var span = findSpan("task:");
        assertEquals("test-agent",    attr(span, "routeiq.agent.id"));
        assertEquals(riq.sessionId,   attr(span, "routeiq.session.id"));
        assertEquals(taskId,          attr(span, "routeiq.task.id"));
        assertEquals("find Paris",    attr(span, "routeiq.task.input_intent"));
        assertEquals("gpt-4o",        attr(span, "routeiq.version.model.name"));
    }

    @Test
    void task_completeSetsSucess() {
        try (var task = riq.task("q")) {
            task.complete(100, null, "test");
        }
        var span = findSpan("task:");
        assertEquals("1",    attr(span, "routeiq.task.completion_status"));
        assertEquals("100",  attr(span, "routeiq.task.total_tokens"));
        assertEquals("test", attr(span, "routeiq.task.cohort"));
    }

    @Test
    void task_failSetsFailure() {
        try (var task = riq.task("q")) {
            task.fail("tool_error");
        }
        var span = findSpan("task:");
        assertEquals("2",          attr(span, "routeiq.task.completion_status"));
        assertEquals("tool_error", attr(span, "routeiq.task.failure_category"));
    }

    @Test
    void task_autoSucceedsOnClose() {
        try (var task = riq.task("q")) { }
        assertEquals("1", attr(findSpan("task:"), "routeiq.task.completion_status"));
    }

    // ── StepHandle ────────────────────────────────────────────────────────

    @Test
    void step_spanNameStartsWithStep() {
        try (var task = riq.task("q")) {
            try (var step = task.step("tool_call", null)) { }
        }
        assertNotNull(findSpan("step:"));
    }

    @Test
    void step_carriesTaskId() {
        String taskId, stepId;
        try (var task = riq.task("q")) {
            taskId = task.taskId;
            try (var step = task.step()) {
                stepId = step.stepId;
            }
        }
        var span = findSpan("step:");
        assertEquals(taskId, attr(span, "routeiq.task.id"));
        assertEquals(stepId, attr(span, "routeiq.step.id"));
    }

    @Test
    void step_indexIncrements() {
        try (var task = riq.task("q")) {
            try (var s = task.step()) { }
            try (var s = task.step()) { }
        }
        var indices = otelTesting.getSpans().stream()
                .filter(s -> s.getName().startsWith("step:"))
                .map(s -> Long.parseLong(attr(s, "routeiq.step.index")))
                .sorted().toList();
        assertEquals(java.util.List.of(1L, 2L), indices);
    }

    // ── ToolHandle ────────────────────────────────────────────────────────

    @Test
    void tool_spanNameIsToolName() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                try (var tool = step.tool("search", Map.of("query", "Paris"), "READ_ONLY")) { }
            }
        }
        assertNotNull(findSpanNamed("tool:search"));
    }

    @Test
    void tool_successSetsStatus1() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                try (var tool = step.tool("search")) {
                    tool.success(50.0);
                }
            }
        }
        var span = findSpanNamed("tool:search");
        assertEquals("1", attr(span, "routeiq.tool.result_status"));
    }

    @Test
    void tool_failSetsStatus2() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                try (var tool = step.tool("search")) {
                    tool.fail("TIMEOUT");
                }
            }
        }
        var span = findSpanNamed("tool:search");
        assertEquals("2",       attr(span, "routeiq.tool.result_status"));
        assertEquals("TIMEOUT", attr(span, "routeiq.tool.error_code"));
    }

    @Test
    void tool_autoSucceedsOnClose() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                try (var tool = step.tool("search")) { }
            }
        }
        assertEquals("1", attr(findSpanNamed("tool:search"), "routeiq.tool.result_status"));
    }

    @Test
    void tool_argsHashIs16Chars() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                try (var tool = step.tool("search", Map.of("query", "Paris"), "READ_ONLY")) { }
            }
        }
        var hash = attr(findSpanNamed("tool:search"), "routeiq.tool.arguments_hash");
        assertNotNull(hash);
        assertEquals(16, hash.length());
    }

    @Test
    void tool_permissionLevelMaps() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                try (var tool = step.tool("write_file", null, "READ_WRITE")) { }
            }
        }
        assertEquals("2", attr(findSpanNamed("tool:write_file"), "routeiq.tool.permission_level"));
    }

    @Test
    void sessionId_consistentAcrossSpans() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                try (var tool = step.tool("search")) { }
            }
        }
        var ids = otelTesting.getSpans().stream()
                .map(s -> attr(s, "routeiq.session.id"))
                .distinct().toList();
        assertEquals(1, ids.size());
        assertEquals(riq.sessionId, ids.get(0));
    }

    // ── v0.3.0 signals ────────────────────────────────────────────────────────

    @Test
    void tool_retryCount() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                try (var tool = step.tool("db_query")) {
                    tool.failRetry("TIMEOUT", 3);
                }
            }
        }
        var span = findSpanNamed("tool:db_query");
        assertEquals("3", attr(span, "routeiq.tool.retry_count"));
    }

    @Test
    void tool_tokenSplit() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                try (var tool = step.tool("llm")) {
                    tool.successTokens(100, 200);
                }
            }
        }
        var span = findSpanNamed("tool:llm");
        assertEquals("100", attr(span, "routeiq.tool.tokens_in"));
        assertEquals("200", attr(span, "routeiq.tool.tokens_out"));
    }

    @Test
    void task_sameToolCount() {
        try (var task = riq.task("q")) {
            for (int i = 0; i < 4; i++) {
                try (var step = task.step()) {
                    try (var tool = step.tool("search")) { }
                }
            }
            task.complete();
        }
        var span = findSpan("task:");
        assertEquals("4", attr(span, "routeiq.same_tool_count"));
    }

    @Test
    void task_sameToolCountNotEmittedForDistinct() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                try (var tool = step.tool("search")) { }
            }
            try (var step = task.step()) {
                try (var tool = step.tool("write")) { }
            }
            task.complete();
        }
        var span = findSpan("task:");
        assertNull(attr(span, "routeiq.same_tool_count"));
    }

    @Test
    void task_escalation() {
        try (var task = riq.task("refund")) {
            task.escalate("amount_too_large", "human_review");
        }
        var span = findSpan("escalation:");
        assertEquals("true",          attr(span, "routeiq.escalation.triggered"));
        assertEquals("amount_too_large", attr(span, "routeiq.escalation.reason"));
        assertEquals("human_review",  attr(span, "routeiq.escalation.target"));
    }

    @Test
    void step_guardrail() {
        try (var task = riq.task("q")) {
            try (var step = task.step()) {
                step.guardrail("pii_filter", true);
            }
        }
        var span = findSpan("guardrail:");
        assertEquals("pii_filter", attr(span, "routeiq.guardrail.type"));
        assertEquals("true",       attr(span, "routeiq.guardrail.blocked"));
    }

    @Test
    void step_replan() {
        try (var task = riq.task("q")) {
            try (var step = task.step("search", null)) {
                step.replan("search_failed_switching_to_cache");
            }
        }
        var span = findSpan("step:");
        assertEquals("true",                             attr(span, "routeiq.replan.triggered"));
        assertEquals("search_failed_switching_to_cache", attr(span, "routeiq.replan.reason"));
    }

    @Test
    void step_modelOverride() {
        try (var task = riq.task("q")) {
            try (var step = task.step(null, null, "claude-opus-4-5")) { }
        }
        var span = findSpan("step:");
        assertEquals("claude-opus-4-5", attr(span, "routeiq.step.model"));
    }

    @Test
    void task_tokenSplitAutoSums() {
        try (var task = riq.task("q")) {
            task.complete(300, 700);
        }
        var span = findSpan("task:");
        assertEquals("300",  attr(span, "routeiq.task.tokens_in"));
        assertEquals("700",  attr(span, "routeiq.task.tokens_out"));
        assertEquals("1000", attr(span, "routeiq.task.total_tokens"));
    }
}

File: sdk/go/routeiq/handles_test.go
package routeiq

import (
	"context"
	"strings"
	"sync"
	"testing"

	"go.opentelemetry.io/otel/attribute"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

// ── test helpers ──────────────────────────────────────────────────────────────

type spanRecorder struct {
	mu    sync.Mutex
	spans []sdktrace.ReadOnlySpan
}

func (r *spanRecorder) OnStart(_ context.Context, _ sdktrace.ReadWriteSpan) {}
func (r *spanRecorder) OnEnd(s sdktrace.ReadOnlySpan) {
	r.mu.Lock()
	r.spans = append(r.spans, s)
	r.mu.Unlock()
}
func (r *spanRecorder) Shutdown(_ context.Context) error  { return nil }
func (r *spanRecorder) ForceFlush(_ context.Context) error { return nil }

func makeTestRiq(t *testing.T) (*RouteIQ, *spanRecorder) {
	t.Helper()
	rec := &spanRecorder{}
	provider := sdktrace.NewTracerProvider(sdktrace.WithSpanProcessor(rec))
	riq := newForTest(provider, Options{
		AgentID:      "test-agent",
		TenantID:     "test-tenant",
		Environment:  "test",
		Model:        "gpt-4o",
		AgentVersion: "1.0.0",
	})
	return riq, rec
}

func (r *spanRecorder) all() []sdktrace.ReadOnlySpan {
	r.mu.Lock()
	defer r.mu.Unlock()
	out := make([]sdktrace.ReadOnlySpan, len(r.spans))
	copy(out, r.spans)
	return out
}

func byName(spans []sdktrace.ReadOnlySpan, prefix string) sdktrace.ReadOnlySpan {
	for _, s := range spans {
		if strings.HasPrefix(s.Name(), prefix) {
			return s
		}
	}
	return nil
}

func attrStr(s sdktrace.ReadOnlySpan, key string) string {
	for _, kv := range s.Attributes() {
		if string(kv.Key) == key {
			return kv.Value.Emit()
		}
	}
	return ""
}

func attrInt(s sdktrace.ReadOnlySpan, key attribute.Key) int64 {
	for _, kv := range s.Attributes() {
		if kv.Key == key {
			return kv.Value.AsInt64()
		}
	}
	return 0
}

// ── TaskHandle ────────────────────────────────────────────────────────────────

func TestTaskSpanName(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "find Paris")
	task.End()

	if byName(rec.all(), "task:") == nil {
		t.Fatal("expected span with name starting task:")
	}
}

func TestTaskEnvelopeAttrs(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "find Paris")
	taskID := task.taskID
	task.End()

	span := byName(rec.all(), "task:")
	if span == nil {
		t.Fatal("task span not found")
	}
	if attrStr(span, "routeiq.agent.id") != "test-agent" {
		t.Error("missing agent.id")
	}
	if attrStr(span, "routeiq.session.id") != riq.sessionID {
		t.Error("session_id mismatch")
	}
	if attrStr(span, "routeiq.task.id") != taskID {
		t.Error("task_id mismatch")
	}
	if attrStr(span, "routeiq.task.input_intent") != "find Paris" {
		t.Error("intent mismatch")
	}
	if attrStr(span, "routeiq.version.model.name") != "gpt-4o" {
		t.Error("model mismatch")
	}
}

func TestTaskComplete(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	task.Complete(WithTokens(100), WithCohort("test"))
	task.End()

	span := byName(rec.all(), "task:")
	if span == nil {
		t.Fatal("task span not found")
	}
	if attrStr(span, "routeiq.task.completion_status") != "1" {
		t.Error("expected success status")
	}
	if attrInt(span, "routeiq.task.total_tokens") != 100 {
		t.Error("tokens mismatch")
	}
	if attrStr(span, "routeiq.task.cohort") != "test" {
		t.Error("cohort mismatch")
	}
}

func TestTaskFail(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	task.Fail("tool_error")
	task.End()

	span := byName(rec.all(), "task:")
	if span == nil {
		t.Fatal("task span not found")
	}
	if attrStr(span, "routeiq.task.completion_status") != "2" {
		t.Error("expected failure status")
	}
	if attrStr(span, "routeiq.task.failure_category") != "tool_error" {
		t.Error("failure_category mismatch")
	}
}

func TestTaskAutoComplete(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	task.End()

	span := byName(rec.all(), "task:")
	if span == nil {
		t.Fatal("task span not found")
	}
	if attrStr(span, "routeiq.task.completion_status") != "1" {
		t.Error("expected auto-success on End()")
	}
}

// ── StepHandle ────────────────────────────────────────────────────────────────

func TestStepSpanName(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step(WithAction("tool_call"))
	step.End()
	task.End()

	if byName(rec.all(), "step:") == nil {
		t.Fatal("expected span with name starting step:")
	}
}

func TestStepCarriesTaskID(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step()
	stepID := step.stepID
	step.End()
	task.End()

	span := byName(rec.all(), "step:")
	if span == nil {
		t.Fatal("step span not found")
	}
	if attrStr(span, "routeiq.task.id") != task.taskID {
		t.Error("task_id mismatch on step span")
	}
	if attrStr(span, "routeiq.step.id") != stepID {
		t.Error("step_id mismatch")
	}
}

func TestStepIndexIncrements(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	s1 := task.Step()
	s1.End()
	s2 := task.Step()
	s2.End()
	task.End()

	var indices []int64
	for _, s := range rec.all() {
		if strings.HasPrefix(s.Name(), "step:") {
			indices = append(indices, attrInt(s, "routeiq.step.index"))
		}
	}
	if len(indices) != 2 || indices[0]+indices[1] != 3 {
		t.Errorf("expected step indices {1,2}, got %v", indices)
	}
}

// ── ToolHandle ────────────────────────────────────────────────────────────────

func TestToolSpanName(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step()
	tool := step.Tool("search", WithArgs(map[string]any{"query": "Paris"}))
	tool.End()
	step.End()
	task.End()

	if byName(rec.all(), "tool:search") == nil {
		t.Fatal("expected span tool:search")
	}
}

func TestToolSuccess(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step()
	tool := step.Tool("search")
	tool.Success(WithLatencyMs(50.0))
	tool.End()
	step.End()
	task.End()

	span := byName(rec.all(), "tool:search")
	if span == nil {
		t.Fatal("tool span not found")
	}
	if attrStr(span, "routeiq.tool.result_status") != "1" {
		t.Error("expected success status")
	}
	if attrStr(span, "routeiq.tool.latency_ms") == "" {
		t.Error("expected latency_ms to be set")
	}
}

func TestToolFail(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step()
	tool := step.Tool("search")
	tool.Fail("TIMEOUT")
	tool.End()
	step.End()
	task.End()

	span := byName(rec.all(), "tool:search")
	if span == nil {
		t.Fatal("tool span not found")
	}
	if attrStr(span, "routeiq.tool.result_status") != "2" {
		t.Error("expected failure status")
	}
	if attrStr(span, "routeiq.tool.error_code") != "TIMEOUT" {
		t.Error("error_code mismatch")
	}
}

func TestToolAutoSucceeds(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step()
	tool := step.Tool("search")
	tool.End()
	step.End()
	task.End()

	span := byName(rec.all(), "tool:search")
	if span == nil {
		t.Fatal("tool span not found")
	}
	if attrStr(span, "routeiq.tool.result_status") != "1" {
		t.Error("expected auto-success on End()")
	}
}

func TestToolArgsHash(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step()
	tool := step.Tool("search", WithArgs(map[string]any{"query": "Paris"}))
	tool.End()
	step.End()
	task.End()

	span := byName(rec.all(), "tool:search")
	if span == nil {
		t.Fatal("tool span not found")
	}
	h := attrStr(span, "routeiq.tool.arguments_hash")
	if len(h) != 16 {
		t.Errorf("expected 16-char hash, got %q", h)
	}
}

func attrFloat64(s sdktrace.ReadOnlySpan, key string) float64 {
	for _, kv := range s.Attributes() {
		if string(kv.Key) == key {
			return kv.Value.AsFloat64()
		}
	}
	return 0
}

func TestSessionIDConsistent(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step()
	tool := step.Tool("search")
	tool.End()
	step.End()
	task.End()

	seen := map[string]bool{}
	for _, s := range rec.all() {
		seen[attrStr(s, "routeiq.session.id")] = true
	}
	if len(seen) != 1 || !seen[riq.sessionID] {
		t.Errorf("session_id inconsistent across spans: %v", seen)
	}
}

// ── v0.3.0 signals ────────────────────────────────────────────────────────────

func TestToolRetryCount(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step()
	tool := step.Tool("db_query")
	tool.Fail("TIMEOUT", WithRetryCount(3))
	tool.End()
	step.End()
	task.End()

	span := byName(rec.all(), "tool:db_query")
	if span == nil {
		t.Fatal("tool span not found")
	}
	if attrInt(span, "routeiq.tool.retry_count") != 3 {
		t.Errorf("retry_count mismatch, got %d", attrInt(span, "routeiq.tool.retry_count"))
	}
}

func TestToolTokenSplit(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step()
	tool := step.Tool("llm")
	tool.Success(WithToolTokensIn(100), WithToolTokensOut(200))
	tool.End()
	step.End()
	task.End()

	span := byName(rec.all(), "tool:llm")
	if span == nil {
		t.Fatal("tool span not found")
	}
	if attrInt(span, "routeiq.tool.tokens_in") != 100 {
		t.Error("tokens_in mismatch")
	}
	if attrInt(span, "routeiq.tool.tokens_out") != 200 {
		t.Error("tokens_out mismatch")
	}
}

func TestSameToolCount(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	for i := 0; i < 4; i++ {
		step := task.Step()
		tool := step.Tool("search")
		tool.End()
		step.End()
	}
	task.Complete()
	task.End()

	span := byName(rec.all(), "task:")
	if span == nil {
		t.Fatal("task span not found")
	}
	if attrInt(span, "routeiq.same_tool_count") != 4 {
		t.Errorf("same_tool_count mismatch, got %d", attrInt(span, "routeiq.same_tool_count"))
	}
}

func TestSameToolCountNotEmittedForDistinct(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	s1 := task.Step()
	s1.Tool("search").End()
	s1.End()
	s2 := task.Step()
	s2.Tool("write").End()
	s2.End()
	task.Complete()
	task.End()

	span := byName(rec.all(), "task:")
	if span == nil {
		t.Fatal("task span not found")
	}
	for _, kv := range span.Attributes() {
		if string(kv.Key) == "routeiq.same_tool_count" {
			t.Error("same_tool_count should not be emitted for distinct tools")
		}
	}
}

func TestEscalation(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "refund")
	task.Escalate("amount_too_large", "human_review")
	task.End()

	span := byName(rec.all(), "escalation:")
	if span == nil {
		t.Fatal("escalation span not found")
	}
	if attrStr(span, "routeiq.escalation.triggered") != "true" {
		t.Error("escalation.triggered mismatch")
	}
	if attrStr(span, "routeiq.escalation.reason") != "amount_too_large" {
		t.Error("escalation.reason mismatch")
	}
	if attrStr(span, "routeiq.escalation.target") != "human_review" {
		t.Error("escalation.target mismatch")
	}
}

func TestGuardrail(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step()
	step.Guardrail("pii_filter", true)
	step.End()
	task.End()

	span := byName(rec.all(), "guardrail:")
	if span == nil {
		t.Fatal("guardrail span not found")
	}
	if attrStr(span, "routeiq.guardrail.type") != "pii_filter" {
		t.Error("guardrail.type mismatch")
	}
	if attrStr(span, "routeiq.guardrail.blocked") != "true" {
		t.Error("guardrail.blocked mismatch")
	}
}

func TestReplan(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step(WithAction("search"))
	step.Replan("search_failed_switching_to_cache")
	step.End()
	task.End()

	span := byName(rec.all(), "step:")
	if span == nil {
		t.Fatal("step span not found")
	}
	if attrStr(span, "routeiq.replan.triggered") != "true" {
		t.Error("replan.triggered mismatch")
	}
	if attrStr(span, "routeiq.replan.reason") != "search_failed_switching_to_cache" {
		t.Error("replan.reason mismatch")
	}
}

func TestStepModelOverride(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	step := task.Step(WithModel("claude-opus-4-5"))
	step.End()
	task.End()

	span := byName(rec.all(), "step:")
	if span == nil {
		t.Fatal("step span not found")
	}
	if attrStr(span, "routeiq.step.model") != "claude-opus-4-5" {
		t.Errorf("step.model mismatch, got %q", attrStr(span, "routeiq.step.model"))
	}
}

func TestTaskTokenSplit(t *testing.T) {
	riq, rec := makeTestRiq(t)
	task := riq.Task(context.Background(), "q")
	task.Complete(WithTokensIn(300), WithTokensOut(700))
	task.End()

	span := byName(rec.all(), "task:")
	if span == nil {
		t.Fatal("task span not found")
	}
	if attrInt(span, "routeiq.task.tokens_in") != 300 {
		t.Error("tokens_in mismatch")
	}
	if attrInt(span, "routeiq.task.tokens_out") != 700 {
		t.Error("tokens_out mismatch")
	}
	if attrInt(span, "routeiq.task.total_tokens") != 1000 {
		t.Error("total_tokens mismatch")
	}
}

func TestSystemIdAndSloTargets(t *testing.T) {
	rec := &spanRecorder{}
	provider := sdktrace.NewTracerProvider(sdktrace.WithSpanProcessor(rec))
	sloSuccess := 0.95
	sloP95 := 2000.0
	riq := newForTest(provider, Options{
		AgentID:          "test-agent",
		TenantID:         "test-tenant",
		Environment:      "test",
		Model:            "gpt-4o",
		AgentVersion:     "1.0.0",
		SystemID:         "checkout-bot",
		UserID:           "user_42",
		SLOSuccessTarget: &sloSuccess,
		SLOP95MsTarget:   &sloP95,
	})
	task := riq.Task(context.Background(), "q")
	task.End()

	span := byName(rec.all(), "task:")
	if span == nil {
		t.Fatal("task span not found")
	}
	if attrStr(span, "routeiq.system.id") != "checkout-bot" {
		t.Errorf("system.id mismatch, got %q", attrStr(span, "routeiq.system.id"))
	}
	if attrStr(span, "routeiq.user.id") != "user_42" {
		t.Errorf("user.id mismatch, got %q", attrStr(span, "routeiq.user.id"))
	}
	if attrFloat64(span, "routeiq.slo.success_target") != 0.95 {
		t.Errorf("slo.success_target mismatch, got %v", attrFloat64(span, "routeiq.slo.success_target"))
	}
	if attrFloat64(span, "routeiq.slo.p95_ms_target") != 2000.0 {
		t.Errorf("slo.p95_ms_target mismatch, got %v", attrFloat64(span, "routeiq.slo.p95_ms_target"))
	}
}
